File: 2026/07-July/01-Wednesday/solution.py
from collections import deque
import heapq

class Solution:
    def maximumSafenessFactor(self, grid: list[list[int]]) -> int:
        n = len(grid)
        
        # If the start or end contains a thief, the safeness factor is instantly 0
        if grid[0][0] == 1 or grid[n-1][n-1] == 1:
            return 0
            
        # Step 1: Multi-source BFS to calculate minimum distance to any thief
        dist_to_thief = [[-1] * n for _ in range(n)]
        q = deque()
        
        for r in range(n):
            for c in range(n):
                if grid[r][c] == 1:
                    q.append((r, c))
                    dist_to_thief[r][c] = 0
                    
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        
        while q:
            r, c = q.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < n and 0 <= nc < n and dist_to_thief[nr][nc] == -1:
                    dist_to_thief[nr][nc] = dist_to_thief[r][c] + 1
                    q.append((nr, nc))
                    
        # Step 2: Modified Dijkstra using Max-Heap to find maximum safeness path
        # Max-heap elements: (-safeness_factor_of_path, r, c)
        max_heap = [(-dist_to_thief[0][0], 0, 0)]
        visited = [[False] * n for _ in range(n)]
        visited[0][0] = True
        
        while max_heap:
            current_safeness, r, c = heapq.heappop(max_heap)
            current_safeness = -current_safeness  # Convert back to positive
            
            # Reached destination
            if r == n - 1 and c == n - 1:
                return current_safeness
                
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < n and 0 <= nc < n and not visited[nr][nc]:
                    visited[nr][nc] = True
                    # The safeness of the path to the neighbor is the bottleneck min
                    next_safeness = min(current_safeness, dist_to_thief[nr][nc])
                    heapq.heappush(max_heap, (-next_safeness, nr, nc))
                    
        return 0

File: 2026/07-July/01-Wednesday/solution.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maximumSafenessFactor(vector<vector<int>>& grid) {
        int n = grid.size();
        
        // If start or end contains a thief, return 0
        if (grid[0][0] == 1 || grid[n-1][n-1] == 1) {
            return 0;
        }
        
        // Step 1: Multi-source BFS
        vector<vector<int>> distToThief(n, vector<int>(n, -1));
        queue<pair<int, int>> q;
        
        for (int r = 0; r < n; ++r) {
            for (int c = 0; c < n; ++c) {
                if (grid[r][c] == 1) {
                    q.push({r, c});
                    distToThief[r][c] = 0;
                }
            }
        }
        
        int directions[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        
        while (!q.empty()) {
            auto [r, c] = q.front();
            q.pop();
            
            for (auto& dir : directions) {
                int nr = r + dir[0];
                int nc = c + dir[1];
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && distToThief[nr][nc] == -1) {
                    distToThief[nr][nc] = distToThief[r][c] + 1;
                    q.push({nr, nc});
                }
            }
        }
        
        // Step 2: Modified Dijkstra using Max-Heap
        // Priority Queue stores: {safeness_factor, {r, c}}
        priority_queue<pair<int, pair<int, int>>> maxHeap;
        vector<vector<bool>> visited(n, vector<bool>(n, false));
        
        maxHeap.push({distToThief[0][0], {0, 0}});
        visited[0][0] = true;
        
        while (!maxHeap.empty()) {
            auto [currentSafeness, cell] = maxHeap.top();
            maxHeap.pop();
            
            int r = cell.first;
            int c = cell.second;
            
            // Reached destination
            if (r == n - 1 && c == n - 1) {
                return currentSafeness;
            }
            
            for (auto& dir : directions) {
                int nr = r + dir[0];
                int nc = c + dir[1];
                
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && !visited[nr][nc]) {
                    visited[nr][nc] = true;
                    int nextSafeness = min(currentSafeness, distToThief[nr][nc]);
                    maxHeap.push({nextSafeness, {nr, nc}});
                }
            }
        }
        
        return 0;
    }
};
